<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>biocmask Usage Guide • biocmask</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="biocmask Usage Guide">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">biocmask</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.12</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/biocmask.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jtlandis/biocmask/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>biocmask Usage Guide</h1>
                        <h4 data-toc-skip class="author">Justin
Landis</h4>
                                    <h4 data-toc-skip class="author">Michael
Love</h4>
                        
            <h4 data-toc-skip class="date">09/24/2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jtlandis/biocmask/blob/main/vignettes/biocmask.Rmd" class="external-link"><code>vignettes/biocmask.Rmd</code></a></small>
      <div class="d-none name"><code>biocmask.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p><code>biocmask</code> provides efficient abstractions to the
<em>SummarizedExperiment</em> such that using common dplyr functions
feels as natural to operating on a <em>data.frame</em> or
<em>tibble</em>. <code>biocmask</code> uses <a href="https://rlang.r-lib.org/reference/topic-data-mask-programming.html" class="external-link">data-masking</a>
from the <code>rlang</code> package in order to connect dplyr functions
to <em>SummarizedExperiment</em> slots in a manner that aims to be
intuitive and avoiding ambiguity in outcomes.</p>
<div class="section level3">
<h3 id="supported-data-types-and-operations">Supported data types and operations<a class="anchor" aria-label="anchor" href="#supported-data-types-and-operations"></a>
</h3>
<p><code>biocmask</code> works on <em>SummarizedExperiment</em> objects,
as well as most classes derived from this, including
<em>DESeqDataSet</em>, <em>SingleCellExperiment</em>, etc.</p>
<p>It supports the following operations:</p>
<ul>
<li><code>mutate</code></li>
<li><code>select</code></li>
<li><code>summarize</code></li>
<li><code>pull</code></li>
<li><code>filter</code></li>
<li><code>arrange</code></li>
</ul>
</div>
<div class="section level3">
<h3 id="typical-use-case">Typical use case<a class="anchor" aria-label="anchor" href="#typical-use-case"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jtlandis/biocmask" class="external-link">biocmask</a></span><span class="op">)</span></span>
<span><span class="co"># add data (mutate) to any of the three tables,</span></span>
<span><span class="co"># assay, colData, rowData,</span></span>
<span><span class="co"># ...using contextual helpers cols() and rows()</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>log_counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="../reference/biocmask-context.html">cols</a></span><span class="op">(</span>treated <span class="op">=</span> <span class="va">dex</span> <span class="op">==</span> <span class="st">"trt"</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="../reference/biocmask-context.html">rows</a></span><span class="op">(</span>new_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene-"</span>, <span class="va">gene_name</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A RangedSummarizedExperiment-tibble Abstraction: 63,677 × 8</span></span></span>
<span><span class="co">##     .features      .samples <span style="color: #949494;">|</span> counts log_counts <span style="color: #949494;">|</span> gene_id gene_name gene_biotype</span></span>
<span><span class="co">##     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494;">|</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494;">|</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       </span></span>
<span><span class="co">## <span style="color: #949494;">  1</span> ENSG000000000… SRR1039… <span style="color: #949494;">|</span>    679       6.52 <span style="color: #949494;">|</span> ENSG00… TSPAN6    protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">  2</span> ENSG000000000… SRR1039… <span style="color: #949494;">|</span>      0       0    <span style="color: #949494;">|</span> ENSG00… TNMD      protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">  3</span> ENSG000000004… SRR1039… <span style="color: #949494;">|</span>    467       6.15 <span style="color: #949494;">|</span> ENSG00… DPM1      protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">  4</span> ENSG000000004… SRR1039… <span style="color: #949494;">|</span>    260       5.56 <span style="color: #949494;">|</span> ENSG00… SCYL3     protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">  5</span> ENSG000000004… SRR1039… <span style="color: #949494;">|</span>     60       4.11 <span style="color: #949494;">|</span> ENSG00… C1orf112  protein_cod…</span></span>
<span><span class="co">##   <span style="color: #949494;">…</span>        <span style="color: #949494;">…</span>           <span style="color: #949494;">…</span>           <span style="color: #949494;">…</span>         <span style="color: #949494;">…</span>       <span style="color: #949494;">…</span>        <span style="color: #949494;">…</span>           <span style="color: #949494;">…</span>     </span></span>
<span><span class="co">## <span style="color: #949494;">n-4</span> ENSG000002734… SRR1039… <span style="color: #949494;">|</span>      0       0    <span style="color: #949494;">|</span> ENSG00… RP11-180… antisense   </span></span>
<span><span class="co">## <span style="color: #949494;">n-3</span> ENSG000002734… SRR1039… <span style="color: #949494;">|</span>      0       0    <span style="color: #949494;">|</span> ENSG00… TSEN34    protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">n-2</span> ENSG000002734… SRR1039… <span style="color: #949494;">|</span>      0       0    <span style="color: #949494;">|</span> ENSG00… RP11-138… lincRNA     </span></span>
<span><span class="co">## <span style="color: #949494;">n-1</span> ENSG000002734… SRR1039… <span style="color: #949494;">|</span>      0       0    <span style="color: #949494;">|</span> ENSG00… AP000230… lincRNA     </span></span>
<span><span class="co">## <span style="color: #949494;">n  </span> ENSG000002734… SRR1039… <span style="color: #949494;">|</span>      0       0    <span style="color: #949494;">|</span> ENSG00… RP11-80H… lincRNA     </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ n = 509,416</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 7 more variables: new_id &lt;chr&gt;, `` &lt;&gt;, SampleName &lt;fct&gt;, cell &lt;fct&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   dex &lt;fct&gt;, albut &lt;fct&gt;, treated &lt;lgl&gt;</span></span></span></code></pre>
<p>The operations can span contexts, and only the necessary data will be
extracted from each context for computation:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span><span class="op">$</span><span class="va">sizeFactor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">.9</span>, <span class="fl">1.1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># making scaled counts, then computing row means:</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>scaled_counts <span class="op">=</span> <span class="va">counts</span> <span class="op">/</span> <span class="va">.cols</span><span class="op">$</span><span class="va">sizeFactor</span>, <span class="co">#</span></span>
<span>         <span class="fu"><a href="../reference/biocmask-context.html">rows</a></span><span class="op">(</span>ave_scl_cts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">.assays_asis</span><span class="op">$</span><span class="va">scaled_counts</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A RangedSummarizedExperiment-tibble Abstraction: 63,677 × 8</span></span></span>
<span><span class="co">##     .features   .samples <span style="color: #949494;">|</span> counts scaled_counts <span style="color: #949494;">|</span> gene_id gene_name gene_biotype</span></span>
<span><span class="co">##     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494;">|</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494;">|</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       </span></span>
<span><span class="co">## <span style="color: #949494;">  1</span> ENSG000000… SRR1039… <span style="color: #949494;">|</span>    679         741.  <span style="color: #949494;">|</span> ENSG00… TSPAN6    protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">  2</span> ENSG000000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… TNMD      protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">  3</span> ENSG000000… SRR1039… <span style="color: #949494;">|</span>    467         510.  <span style="color: #949494;">|</span> ENSG00… DPM1      protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">  4</span> ENSG000000… SRR1039… <span style="color: #949494;">|</span>    260         284.  <span style="color: #949494;">|</span> ENSG00… SCYL3     protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">  5</span> ENSG000000… SRR1039… <span style="color: #949494;">|</span>     60          65.5 <span style="color: #949494;">|</span> ENSG00… C1orf112  protein_cod…</span></span>
<span><span class="co">##   <span style="color: #949494;">…</span>      <span style="color: #949494;">…</span>          <span style="color: #949494;">…</span>           <span style="color: #949494;">…</span>            <span style="color: #949494;">…</span>       <span style="color: #949494;">…</span>        <span style="color: #949494;">…</span>           <span style="color: #949494;">…</span>     </span></span>
<span><span class="co">## <span style="color: #949494;">n-4</span> ENSG000002… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… RP11-180… antisense   </span></span>
<span><span class="co">## <span style="color: #949494;">n-3</span> ENSG000002… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… TSEN34    protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">n-2</span> ENSG000002… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… RP11-138… lincRNA     </span></span>
<span><span class="co">## <span style="color: #949494;">n-1</span> ENSG000002… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… AP000230… lincRNA     </span></span>
<span><span class="co">## <span style="color: #949494;">n  </span> ENSG000002… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… RP11-80H… lincRNA     </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ n = 509,416</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 7 more variables: ave_scl_cts &lt;dbl&gt;, `` &lt;&gt;, SampleName &lt;fct&gt;, cell &lt;fct&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   dex &lt;fct&gt;, albut &lt;fct&gt;, sizeFactor &lt;dbl&gt;</span></span></span></code></pre>
<p>Calling <code>.cols</code> in the assay context produces an object of
the matching size and orientation to the other assay data.</p>
<p>Alternatively we could have used <em>purrr</em> to compute row
means:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>scaled_counts <span class="op">=</span> <span class="va">counts</span> <span class="op">/</span> <span class="va">.cols</span><span class="op">$</span><span class="va">sizeFactor</span>,</span>
<span>         <span class="co"># You may expect a list when accessing other contexts</span></span>
<span>         <span class="co"># from either the rows() or cols() contexts.</span></span>
<span>         <span class="fu"><a href="../reference/biocmask-context.html">rows</a></span><span class="op">(</span>ave_scl_cts <span class="op">=</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_dbl</a></span><span class="op">(</span><span class="va">.assays</span><span class="op">$</span><span class="va">scaled_counts</span>, <span class="va">mean</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A RangedSummarizedExperiment-tibble Abstraction: 63,677 × 8</span></span></span>
<span><span class="co">##     .features   .samples <span style="color: #949494;">|</span> counts scaled_counts <span style="color: #949494;">|</span> gene_id gene_name gene_biotype</span></span>
<span><span class="co">##     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494;">|</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494;">|</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       </span></span>
<span><span class="co">## <span style="color: #949494;">  1</span> ENSG000000… SRR1039… <span style="color: #949494;">|</span>    679         741.  <span style="color: #949494;">|</span> ENSG00… TSPAN6    protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">  2</span> ENSG000000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… TNMD      protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">  3</span> ENSG000000… SRR1039… <span style="color: #949494;">|</span>    467         510.  <span style="color: #949494;">|</span> ENSG00… DPM1      protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">  4</span> ENSG000000… SRR1039… <span style="color: #949494;">|</span>    260         284.  <span style="color: #949494;">|</span> ENSG00… SCYL3     protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">  5</span> ENSG000000… SRR1039… <span style="color: #949494;">|</span>     60          65.5 <span style="color: #949494;">|</span> ENSG00… C1orf112  protein_cod…</span></span>
<span><span class="co">##   <span style="color: #949494;">…</span>      <span style="color: #949494;">…</span>          <span style="color: #949494;">…</span>           <span style="color: #949494;">…</span>            <span style="color: #949494;">…</span>       <span style="color: #949494;">…</span>        <span style="color: #949494;">…</span>           <span style="color: #949494;">…</span>     </span></span>
<span><span class="co">## <span style="color: #949494;">n-4</span> ENSG000002… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… RP11-180… antisense   </span></span>
<span><span class="co">## <span style="color: #949494;">n-3</span> ENSG000002… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… TSEN34    protein_cod…</span></span>
<span><span class="co">## <span style="color: #949494;">n-2</span> ENSG000002… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… RP11-138… lincRNA     </span></span>
<span><span class="co">## <span style="color: #949494;">n-1</span> ENSG000002… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… AP000230… lincRNA     </span></span>
<span><span class="co">## <span style="color: #949494;">n  </span> ENSG000002… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… RP11-80H… lincRNA     </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ n = 509,416</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 7 more variables: ave_scl_cts &lt;dbl&gt;, `` &lt;&gt;, SampleName &lt;fct&gt;, cell &lt;fct&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   dex &lt;fct&gt;, albut &lt;fct&gt;, sizeFactor &lt;dbl&gt;</span></span></span></code></pre>
<p>See below for details on how objects are made available across
contexts.</p>
<p><code>biocmask</code> also enables common grouping and summarization
routines:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summary</span> <span class="op">&lt;-</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="fu"><a href="../reference/biocmask-context.html">rows</a></span><span class="op">(</span><span class="va">gene_biotype</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>col_sums <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>            <span class="co"># may rename rows with .features</span></span>
<span>            <span class="fu"><a href="../reference/biocmask-context.html">rows</a></span><span class="op">(</span>.features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gene_biotype</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># summarize returns a SummarizedExperiment here,</span></span>
<span><span class="co"># retaining rowData and colData</span></span>
<span></span>
<span><span class="va">summary</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 30 rows and 1 column</span></span>
<span><span class="co">##                                    gene_biotype</span></span>
<span><span class="co">##                                     &lt;character&gt;</span></span>
<span><span class="co">## protein_coding                   protein_coding</span></span>
<span><span class="co">## pseudogene                           pseudogene</span></span>
<span><span class="co">## processed_transcript       processed_transcript</span></span>
<span><span class="co">## antisense                             antisense</span></span>
<span><span class="co">## lincRNA                                 lincRNA</span></span>
<span><span class="co">## ...                                         ...</span></span>
<span><span class="co">## IG_C_pseudogene                 IG_C_pseudogene</span></span>
<span><span class="co">## TR_D_gene                             TR_D_gene</span></span>
<span><span class="co">## IG_J_pseudogene                 IG_J_pseudogene</span></span>
<span><span class="co">## 3prime_overlapping_ncrna 3prime_overlapping_n..</span></span>
<span><span class="co">## processed_pseudogene       processed_pseudogene</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualizing the output as a tibble:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="va">summary</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">col_sums</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"type"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 30 × 9</span></span></span>
<span><span class="co">##    type        SRR1039508 SRR1039509 SRR1039512 SRR1039513 SRR1039516 SRR1039517</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> protein_co…   19<span style="text-decoration: underline;">413</span>626      <span style="text-decoration: underline;">45</span>654          4       <span style="text-decoration: underline;">1</span>188      <span style="text-decoration: underline;">96</span>378          0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> pseudogene    17<span style="text-decoration: underline;">741</span>060      <span style="text-decoration: underline;">45</span>864          4        462      <span style="text-decoration: underline;">38</span>656          0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> processed_…   23<span style="text-decoration: underline;">926</span>011     <span style="text-decoration: underline;">133</span>335          0       <span style="text-decoration: underline;">1</span>049      <span style="text-decoration: underline;">64</span>884          0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> antisense     14<span style="text-decoration: underline;">360</span>299     <span style="text-decoration: underline;">120</span>060          4       <span style="text-decoration: underline;">1</span>113      <span style="text-decoration: underline;">36</span>267          0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> lincRNA       23<span style="text-decoration: underline;">003</span>444     <span style="text-decoration: underline;">206</span>075       <span style="text-decoration: underline;">6</span>038        626      <span style="text-decoration: underline;">81</span>606          0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> polymorphi…   29<span style="text-decoration: underline;">233</span>398     <span style="text-decoration: underline;">125</span>015       <span style="text-decoration: underline;">5</span>618        803      <span style="text-decoration: underline;">88</span>868          0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> IG_V_pseud…   18<span style="text-decoration: underline;">114</span>369     <span style="text-decoration: underline;">145</span>078       <span style="text-decoration: underline;">7</span>662        316      <span style="text-decoration: underline;">44</span>385          0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> IG_V_gene     20<span style="text-decoration: underline;">103</span>401     <span style="text-decoration: underline;">170</span>641       <span style="text-decoration: underline;">5</span>579        256      <span style="text-decoration: underline;">92</span>499          0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> sense_over…     <span style="text-decoration: underline;">807</span>285     <span style="text-decoration: underline;">147</span>563       <span style="text-decoration: underline;">7</span>869        339        491          0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> sense_intr…     <span style="text-decoration: underline;">733</span>916     <span style="text-decoration: underline;">149</span>486       <span style="text-decoration: underline;">9</span>443        202        502          0</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 20 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 2 more variables: SRR1039520 &lt;dbl&gt;, SRR1039521 &lt;dbl&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="related-work">Related work<a class="anchor" aria-label="anchor" href="#related-work"></a>
</h3>
<p>We note that <code>biocmask</code> is highly related to other
<em>tidyomics</em> projects including:</p>
<ul>
<li>tidySummarizedExperiment</li>
<li>plyranges</li>
<li>DFplyr</li>
<li>and more.</li>
</ul>
<p>Here we have focused on the design principles of function
<em>endomorphism</em> (returning the same object that was input),
avoiding ambiguity through strictly defined behavior (potentially at the
expense of longer code), and allowing the user the convenience of
multiple expressions for the same result, some of which may have
improved computational performance.</p>
</div>
</div>
<div class="section level2">
<h2 id="manipulating-summarizedexperiment-with-biocmask">Manipulating <em>SummarizedExperiment</em> with
<code>biocmask</code><a class="anchor" aria-label="anchor" href="#manipulating-summarizedexperiment-with-biocmask"></a>
</h2>
<!-- copied from README -->
<!-- unsure if it is better here or in README -->
<p>The <code>SummarizedExperiment</code> object contains three main
components/“contexts” that we mask, the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays()</a></code>,
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code><a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;At this moment &lt;code&gt;rowRanges()&lt;/code&gt; is not supported
in &lt;code&gt;biocmask&lt;/code&gt; but may become its own pronoun in the future.&lt;/p&gt;"><sup>1</sup></a> and <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>.</p>
<div class="float">
<img src="../reference/figures/Overview-bindings.png" alt="Simplified view of data masking structure. Figure made with Biorender"><div class="figcaption">Simplified view of data masking structure.
Figure made with <a href="https://biorender.com" class="external-link">Biorender</a>
</div>
</div>
<p><code>biocmask</code> provides variables as-is to data <strong>within
their current contexts</strong> enabling you to call S4 methods on S4
objects with <code>dplyr</code> verbs. If you require access to
variables <em>outside the context</em>, you may use pronouns made
available through <code>biocmask</code> to specify where to find those
variables.</p>
<div class="float">
<img src="../reference/figures/Overview-pronouns.png" alt="Simplified view of reshaping pronouns. Arrows indicates to where the pronoun provides access. For each pronoun listed, there is an _asis variant that returns underlying data without reshaping it to fit the context. Figure made with Biorender"><div class="figcaption">Simplified view of reshaping pronouns. Arrows
indicates to where the pronoun provides access. For each pronoun listed,
there is an <code>_asis</code> variant that returns underlying data
without reshaping it to fit the context. Figure made with <a href="https://biorender.com" class="external-link">Biorender</a>
</div>
</div>
<p>The <code>.assays</code>, <code>.rows</code> and <code>.cols</code>
pronouns outputs depends on the evaluating context. Users should expect
that the underlying data returned from <code>.rows</code> or
<code>.cols</code> pronouns in the <em><strong>assays
context</strong></em> is a vector, replicated to match size of the assay
context.</p>
<p><br></p>
<p>Alternatively, using a pronoun in either the <code><a href="../reference/biocmask-context.html">rows()</a></code> or
<code><a href="../reference/biocmask-context.html">cols()</a></code> contexts will return a list equal in length to
either <code>nrows(rowData())</code> or <code>nrows(colData())</code>
respectively.</p>
<div class="section level3">
<h3 id="assay-context">assay context<a class="anchor" aria-label="anchor" href="#assay-context"></a>
</h3>
<!-- taken from the presentation -->
<ul>
<li>Default evaluation context</li>
<li>
<code>.assays</code>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>→</mo><annotation encoding="application/x-tex">\to</annotation></semantics></math>
contextual pronoun, returns list of the matrix, sliced by the dimension
it was referenced from.
<ul>
<li>within the rowData context: <code>.assays$foo</code> is an alias for
<code>lapply(seq_len(nrow()), \(i, x) x[i,,drop=FALSE], x = foo)</code>
</li>
<li>within the colData context: <code>.assays$foo</code> is an alias for
<code>lapply(seq_len(ncol()), \(i, x) x[,i,drop=FALSE], x = foo)</code>
</li>
</ul>
</li>
<li>
<code>.assays_asis</code>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>→</mo><annotation encoding="application/x-tex">\to</annotation></semantics></math>
pronoun to direct bindings in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays()</a></code>
</li>
<li>
<code>assay_ctx(expr, asis = FALSE)</code>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>→</mo><annotation encoding="application/x-tex">\to</annotation></semantics></math>
short hand to bind the assay context in front of the current
context.</li>
</ul>
</div>
<div class="section level3">
<h3 id="rows-context">rows context<a class="anchor" aria-label="anchor" href="#rows-context"></a>
</h3>
<ul>
<li>
<code>rows(...)</code>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>→</mo><annotation encoding="application/x-tex">\to</annotation></semantics></math>
sentinel function call to indicate evaluation context.</li>
<li>
<code>.rows</code>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>→</mo><annotation encoding="application/x-tex">\to</annotation></semantics></math>
contextual pronoun
<ul>
<li>within assay context: <code>.rows$foo</code> is an alias for
<code>vctrs::vec_rep(foo, times = ncol())</code>
</li>
<li>within colData context: <code>.rows$foo</code> is an alias for
<code>vctrs::vec_rep(list(foo), times = n())</code>
</li>
</ul>
</li>
<li>
<code>.rows_asis</code>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>→</mo><annotation encoding="application/x-tex">\to</annotation></semantics></math>
pronoun to direct bindings in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>
</li>
<li>
<code>row_ctx(expr, asis = FALSE)</code>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>→</mo><annotation encoding="application/x-tex">\to</annotation></semantics></math>
shorthand to bind the rowData context in front of the current
context</li>
</ul>
</div>
<div class="section level3">
<h3 id="cols-context">cols context<a class="anchor" aria-label="anchor" href="#cols-context"></a>
</h3>
<ul>
<li>
<code>cols(...)</code>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>→</mo><annotation encoding="application/x-tex">\to</annotation></semantics></math>
sentinel function call to indicate evaluation context.</li>
<li>
<code>.cols</code>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>→</mo><annotation encoding="application/x-tex">\to</annotation></semantics></math>
contextual pronoun
<ul>
<li>within assay context: <code>.cols$foo</code> is an alias for
<code>vctrs::vec_rep_each(foo, times = nrow())</code>
</li>
<li>within rowData context: <code>.rows$foo</code> is an alias for
<code>vctrs::vec_rep(list(foo), times = n())</code>
</li>
</ul>
</li>
<li>
<code>.cols_asis</code>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>→</mo><annotation encoding="application/x-tex">\to</annotation></semantics></math>
pronoun to direct bindings in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>
</li>
<li>
<code>col_ctx(expr, asis = FALSE)</code>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>→</mo><annotation encoding="application/x-tex">\to</annotation></semantics></math>
shorthand to bind the colData context in front of the current
context</li>
</ul>
</div>
<div class="section level3">
<h3 id="multiple-expressions-enabled-via-biocmask">Multiple expressions enabled via <code>biocmask</code><a class="anchor" aria-label="anchor" href="#multiple-expressions-enabled-via-biocmask"></a>
</h3>
<p>We can compare two ways of dividing out a vector from
<code>colData</code> along the columns of <code>assay</code> data:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># here the `.cols$` pronoun reshapes the data to fit the `assays` context</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>scaled_counts <span class="op">=</span> <span class="va">counts</span> <span class="op">/</span> <span class="va">.cols</span><span class="op">$</span><span class="va">sizeFactor</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A RangedSummarizedExperiment-tibble Abstraction: 63,677 × 8</span></span></span>
<span><span class="co">##     .features .samples <span style="color: #949494;">|</span> counts scaled_counts <span style="color: #949494;">|</span> gene_id gene_name gene_biotype <span style="color: #949494;">|</span></span></span>
<span><span class="co">##     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494;">|</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494;">|</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  1</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>    679         741.  <span style="color: #949494;">|</span> ENSG00… TSPAN6    protein_cod… <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  2</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… TNMD      protein_cod… <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  3</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>    467         510.  <span style="color: #949494;">|</span> ENSG00… DPM1      protein_cod… <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  4</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>    260         284.  <span style="color: #949494;">|</span> ENSG00… SCYL3     protein_cod… <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  5</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>     60          65.5 <span style="color: #949494;">|</span> ENSG00… C1orf112  protein_cod… <span style="color: #949494;">|</span></span></span>
<span><span class="co">##   <span style="color: #949494;">…</span>     <span style="color: #949494;">…</span>         <span style="color: #949494;">…</span>           <span style="color: #949494;">…</span>            <span style="color: #949494;">…</span>       <span style="color: #949494;">…</span>        <span style="color: #949494;">…</span>           <span style="color: #949494;">…</span>       </span></span>
<span><span class="co">## <span style="color: #949494;">n-4</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… RP11-180… antisense    <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">n-3</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… TSEN34    protein_cod… <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">n-2</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… RP11-138… lincRNA      <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">n-1</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… AP000230… lincRNA      <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">n  </span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… RP11-80H… lincRNA      <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ n = 509,416</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 5 more variables: SampleName &lt;fct&gt;, cell &lt;fct&gt;, dex &lt;fct&gt;, albut &lt;fct&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   sizeFactor &lt;dbl&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this is equivalent to the following, where we have to transpose</span></span>
<span><span class="co"># the `counts` assay data twice to enable the correct recycling</span></span>
<span><span class="co"># of the size factor vector</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>scaled_counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">/</span> <span class="va">.cols_asis</span><span class="op">$</span><span class="va">sizeFactor</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A RangedSummarizedExperiment-tibble Abstraction: 63,677 × 8</span></span></span>
<span><span class="co">##     .features .samples <span style="color: #949494;">|</span> counts scaled_counts <span style="color: #949494;">|</span> gene_id gene_name gene_biotype <span style="color: #949494;">|</span></span></span>
<span><span class="co">##     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494;">|</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494;">|</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  1</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>    679         741.  <span style="color: #949494;">|</span> ENSG00… TSPAN6    protein_cod… <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  2</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… TNMD      protein_cod… <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  3</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>    467         510.  <span style="color: #949494;">|</span> ENSG00… DPM1      protein_cod… <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  4</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>    260         284.  <span style="color: #949494;">|</span> ENSG00… SCYL3     protein_cod… <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  5</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>     60          65.5 <span style="color: #949494;">|</span> ENSG00… C1orf112  protein_cod… <span style="color: #949494;">|</span></span></span>
<span><span class="co">##   <span style="color: #949494;">…</span>     <span style="color: #949494;">…</span>         <span style="color: #949494;">…</span>           <span style="color: #949494;">…</span>            <span style="color: #949494;">…</span>       <span style="color: #949494;">…</span>        <span style="color: #949494;">…</span>           <span style="color: #949494;">…</span>       </span></span>
<span><span class="co">## <span style="color: #949494;">n-4</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… RP11-180… antisense    <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">n-3</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… TSEN34    protein_cod… <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">n-2</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… RP11-138… lincRNA      <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">n-1</span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… AP000230… lincRNA      <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">n  </span> ENSG0000… SRR1039… <span style="color: #949494;">|</span>      0           0   <span style="color: #949494;">|</span> ENSG00… RP11-80H… lincRNA      <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ n = 509,416</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 5 more variables: SampleName &lt;fct&gt;, cell &lt;fct&gt;, dex &lt;fct&gt;, albut &lt;fct&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   sizeFactor &lt;dbl&gt;</span></span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="advanced-features">Advanced features<a class="anchor" aria-label="anchor" href="#advanced-features"></a>
</h2>
<div class="section level3">
<h3 id="object-integrity">Object integrity<a class="anchor" aria-label="anchor" href="#object-integrity"></a>
</h3>
<p><code>biocmask</code> provides an opinionated framework for how
<code>dplyr</code> verbs should interact with
<code>SummarizedExperiment</code> objects. In general,
<code>biocmask</code> will not allow any operations that it could not
guarantee a valid return object.</p>
<p>It is for this reason that <code><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange()</a></code>,
<code><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by()</a></code> do not allow
operations in the default assay context, as this would likely break the
assumed structure of a <code>SummarizedExperiment</code> object.</p>
</div>
<div class="section level3">
<h3 id="group_by">
<code>group_by()</code><a class="anchor" aria-label="anchor" href="#group_by"></a>
</h3>
<p><code>biocmask</code> also supports <code>group_by</code> operations
allowing users to query information with <code><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">dplyr::n()</a></code> or
<code><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">dplyr::cur_group_id()</a></code>. However due to the linked structure
of a <code>SummarizedExperiment</code> object and <code>biocmask</code>
providing multiple evaluation contexts, grouping operations would be
complex and return values would be potentionally ambiguous.</p>
<p>It is for this reason that groupings are themselves
<strong>contextual</strong>. The assay context is dependently linked to
both the groupings of the rows and cols contexts but, the grouping of
rows is ignored when viewed from the cols context and similarly the
grouping of cols is ignored when viewed from the rows context. In this
way, we have chosen to make the groupings of rows and cols independent
between each other. The below figure attempts to show how groupings are
conditionally dropped for the scope of an evaluation.</p>
<div class="float">
<img src="../reference/figures/contextual_groups.png" alt="When evaluating in row context, groupings along colData() are temporarily ignored. Figure created with Biorender"><div class="figcaption">When evaluating in row context, groupings along
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> are temporarily ignored. Figure created with <a href="https://biorender.com" class="external-link">Biorender</a>
</div>
</div>
<p>To further motivate this choice, suppose we did not drop grouping
values. Assume you have a small 5 by 4 <code>SummarizedExperiment</code>
object. Both of the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> and <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>
are grouped such that there are 2 groups in both <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>
and <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> totaling in 4 groups across the assays.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">se_simple</span>, <span class="fu"><a href="../reference/biocmask-context.html">rows</a></span><span class="op">(</span><span class="va">direction</span><span class="op">)</span>, <span class="fu"><a href="../reference/biocmask-context.html">cols</a></span><span class="op">(</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="fu"><a href="../reference/biocmask-context.html">rows</a></span><span class="op">(</span>data_from_cols <span class="op">=</span> <span class="va">.cols</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The above syntax implies we wish to move data from the
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> into the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>. From a
previously established conventions, we would expect the output to be an
alias for
<code>vctrs::vec_rep(list(condition), times = n())</code>.<br>
Additionally the <code><a href="../reference/biocmask-context.html">rows()</a></code> sentinal will expect that the
output of <code>.cols$condition</code> will need to match the size of
the evaluation context.</p>
<p>Unfortunately, this becomes extremely difficult to resolve with the
current conventions. Without dropping the <code><a href="../reference/biocmask-context.html">cols()</a></code> groupings,
each <code><a href="../reference/biocmask-context.html">rows()</a></code> grouping is evaluated equal to the number of
groups in <code><a href="../reference/biocmask-context.html">cols()</a></code>. At first glance, this may seem desirable,
but the problem arises when considering how theses outputs across groups
should be stored per group of <code><a href="../reference/biocmask-context.html">rows()</a></code>. For example, should
the output of the <code>.cols$condition</code> return a list equal to
the number of groups of the column context? If yes, we would need to
consider the size stability of the output context! Assuming that
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> has at least one group with three elements, there
is no guarentee it would fit (this also makes a poor assumption that the
elements of <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> somehow correspond to the groups of
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>). Thus we would be left with wrapping all the
results in a list and replicating to the appropriate size. When its all
said and done, we would have a list of lists, which is difficult to
reason about, potentionally unexpected and painful to work with. For
this reason the only groupings present in the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>
context are the groupings in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>, and similarly for
the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> context.</p>
</div>
<div class="section level3">
<h3 id="printing">Printing<a class="anchor" aria-label="anchor" href="#printing"></a>
</h3>
<p>Motivated by the <code>show</code>/<code>print</code> function in
<em>tidySummarizedExperiment</em>, we visualize the data as if it was
tabular. Also following the example of
<em>tidySummarizedExperiment</em>, <code>biocmask</code> offers the
option to turn off this and return to the default <code>show</code>
method for <em>SummarizedExperiment</em>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"restore_SingleCellExperiment_show"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Since <code>biocmask</code> aims to leave the input data as-is when
possible, we have considered providing support for printing
<code>S4Vectors</code> within a <code><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble()</a></code>. To be clear,
<code>biocmask</code> <strong>will not</strong> allow you to put
<code>S4Vectors</code> inside a <code><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble()</a></code>, but will allow
for <code>S4Vectors</code> to be printed with <code>pillar()</code>, the
formatting engine behind <code><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble()</a></code> pretty printing.</p>
<p>To enable this behavior, before any data is reported to the user, we
proxy any <code>S4Vector</code> with a custom <code>vctrs_vctr</code>
object with <code><a href="../reference/biocmask-printing.html">biocmask::vec_phantom()</a></code>. In truth, the
<code>vec_phantom</code> object is a simple integer vector with a
“phantomData” attribute. This allows us to carry along
<code>S4Vector</code> through <code>pillar()</code>’s printing pipeline
until it is time to print the column.</p>
<p>The <code>pillar_shaft()</code> method for <code>vec_phantom</code>
will format the S4 data with <code><a href="../reference/biocmask-printing.html">biocmask_pillar_format()</a></code>
generic, which by default calls <code><a href="https://rdrr.io/pkg/S4Vectors/man/show-utils.html" class="external-link">S4Vectors::showAsCell()</a></code>.
Users are free to create their own methods for S4 vectors that differ
from <code><a href="https://rdrr.io/pkg/S4Vectors/man/show-utils.html" class="external-link">S4Vectors::showAsCell()</a></code> if they like, as seen in
<code><a href="../reference/biocmask-printing.html">?`biocmask-printing`</a></code></p>
</div>
<div class="section level3">
<h3 id="renaming-rows-or-columns">renaming rows or columns<a class="anchor" aria-label="anchor" href="#renaming-rows-or-columns"></a>
</h3>
<p>Inspired by <code>tidySummarizedExperiment</code>,
<code>biocmask</code> provides access to the rownames and colnames of a
<code>SummarizedExperiment</code> object by installing
<code>.features</code> and <code>.samples</code> into the
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> and <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> contexts respectively.
These are special in that assigning a value to <code>.features</code> in
the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> context or <code>.samples</code> in the
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> context does not create a new column, but changes
the rownames or colnames of the resulting object.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_simple</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SummarizedExperiment-tibble Abstraction: 5 × 4</span></span></span>
<span><span class="co">##     .features .samples <span style="color: #949494;">|</span> counts logcounts <span style="color: #949494;">|</span> gene  length direction <span style="color: #949494;">|</span> sample</span></span>
<span><span class="co">##     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494;">|</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494;">|</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494;">|</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## <span style="color: #949494;">  1</span> row_1     col_1    <span style="color: #949494;">|</span>     14      2.64 <span style="color: #949494;">|</span> g1         1 -         <span style="color: #949494;">|</span> s1    </span></span>
<span><span class="co">## <span style="color: #949494;">  2</span> row_2     col_1    <span style="color: #949494;">|</span>     19      2.94 <span style="color: #949494;">|</span> g2        24 +         <span style="color: #949494;">|</span> s1    </span></span>
<span><span class="co">## <span style="color: #949494;">  3</span> row_3     col_1    <span style="color: #949494;">|</span>     16      2.77 <span style="color: #949494;">|</span> g3        60 +         <span style="color: #949494;">|</span> s1    </span></span>
<span><span class="co">## <span style="color: #949494;">  4</span> row_4     col_1    <span style="color: #949494;">|</span>     11      2.40 <span style="color: #949494;">|</span> g4        39 -         <span style="color: #949494;">|</span> s1    </span></span>
<span><span class="co">## <span style="color: #949494;">  5</span> row_5     col_1    <span style="color: #949494;">|</span>     18      2.89 <span style="color: #949494;">|</span> g5        37 +         <span style="color: #949494;">|</span> s1    </span></span>
<span><span class="co">##   <span style="color: #949494;">…</span>   <span style="color: #949494;">…</span>         <span style="color: #949494;">…</span>             <span style="color: #949494;">…</span>        <span style="color: #949494;">…</span>     <span style="color: #949494;">…</span>         <span style="color: #949494;">…</span> <span style="color: #949494;">…</span>            <span style="color: #949494;">…</span>    </span></span>
<span><span class="co">## <span style="color: #949494;">n-4</span> row_1     col_4    <span style="color: #949494;">|</span>      9      2.20 <span style="color: #949494;">|</span> g1         1 -         <span style="color: #949494;">|</span> s4    </span></span>
<span><span class="co">## <span style="color: #949494;">n-3</span> row_2     col_4    <span style="color: #949494;">|</span>      4      1.39 <span style="color: #949494;">|</span> g2        24 +         <span style="color: #949494;">|</span> s4    </span></span>
<span><span class="co">## <span style="color: #949494;">n-2</span> row_3     col_4    <span style="color: #949494;">|</span>     20      3.00 <span style="color: #949494;">|</span> g3        60 +         <span style="color: #949494;">|</span> s4    </span></span>
<span><span class="co">## <span style="color: #949494;">n-1</span> row_4     col_4    <span style="color: #949494;">|</span>      3      1.10 <span style="color: #949494;">|</span> g4        39 -         <span style="color: #949494;">|</span> s4    </span></span>
<span><span class="co">## <span style="color: #949494;">n  </span> row_5     col_4    <span style="color: #949494;">|</span>      5      1.61 <span style="color: #949494;">|</span> g5        37 +         <span style="color: #949494;">|</span> s4    </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ n = 20</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: condition &lt;chr&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># moving data to rownames and colnames</span></span>
<span><span class="va">se_simple</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    orig_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span></span>
<span>      <span class="st">"%s-%s"</span>,</span>
<span>      <span class="co"># .features is installed in the rows() context</span></span>
<span>      <span class="va">.rows</span><span class="op">$</span><span class="va">.features</span>,</span>
<span>      <span class="co"># .samples is installed in the cols() context</span></span>
<span>      <span class="va">.cols</span><span class="op">$</span><span class="va">.samples</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="../reference/biocmask-context.html">rows</a></span><span class="op">(</span>.features <span class="op">=</span> <span class="va">gene</span>,</span>
<span>         <span class="co"># deleting rowData column</span></span>
<span>         gene <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="../reference/biocmask-context.html">cols</a></span><span class="op">(</span>.samples <span class="op">=</span> <span class="va">sample</span>,</span>
<span>         <span class="co"># deleting colData column</span></span>
<span>         sample <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SummarizedExperiment-tibble Abstraction: 5 × 4</span></span></span>
<span><span class="co">##     .features .samples <span style="color: #949494;">|</span> counts logcounts orig_names  <span style="color: #949494;">|</span> length direction <span style="color: #949494;">|</span></span></span>
<span><span class="co">##     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494;">|</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494;">|</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  1</span> g1        s1       <span style="color: #949494;">|</span>     14      2.64 row_1-col_1 <span style="color: #949494;">|</span>      1 -         <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  2</span> g2        s1       <span style="color: #949494;">|</span>     19      2.94 row_2-col_1 <span style="color: #949494;">|</span>     24 +         <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  3</span> g3        s1       <span style="color: #949494;">|</span>     16      2.77 row_3-col_1 <span style="color: #949494;">|</span>     60 +         <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  4</span> g4        s1       <span style="color: #949494;">|</span>     11      2.40 row_4-col_1 <span style="color: #949494;">|</span>     39 -         <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">  5</span> g5        s1       <span style="color: #949494;">|</span>     18      2.89 row_5-col_1 <span style="color: #949494;">|</span>     37 +         <span style="color: #949494;">|</span></span></span>
<span><span class="co">##   <span style="color: #949494;">…</span>  <span style="color: #949494;">…</span>         <span style="color: #949494;">…</span>              <span style="color: #949494;">…</span>        <span style="color: #949494;">…</span>       <span style="color: #949494;">…</span>             <span style="color: #949494;">…</span> <span style="color: #949494;">…</span>          </span></span>
<span><span class="co">## <span style="color: #949494;">n-4</span> g1        s4       <span style="color: #949494;">|</span>      9      2.20 row_1-col_4 <span style="color: #949494;">|</span>      1 -         <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">n-3</span> g2        s4       <span style="color: #949494;">|</span>      4      1.39 row_2-col_4 <span style="color: #949494;">|</span>     24 +         <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">n-2</span> g3        s4       <span style="color: #949494;">|</span>     20      3.00 row_3-col_4 <span style="color: #949494;">|</span>     60 +         <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">n-1</span> g4        s4       <span style="color: #949494;">|</span>      3      1.10 row_4-col_4 <span style="color: #949494;">|</span>     39 -         <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;">n  </span> g5        s4       <span style="color: #949494;">|</span>      5      1.61 row_5-col_4 <span style="color: #949494;">|</span>     37 +         <span style="color: #949494;">|</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ n = 20</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: condition &lt;chr&gt;</span></span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="troubleshooting-and-best-practices">Troubleshooting and best practices<a class="anchor" aria-label="anchor" href="#troubleshooting-and-best-practices"></a>
</h2>
<p>while <code>biocmask</code> takes inspiration from the data masks
used in <code>dplyr</code>, they are unfortunately more complex. This
means there is some overhead in creating the evaluation mask per dplyr
verb. Try to reduce the number of <code>dplyr</code> verb calls and
instead increase the content of each verb. For example instead of
doing:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">.data</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>foo <span class="op">=</span> <span class="va">bar</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>baz <span class="op">=</span> <span class="va">foo</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>do the following</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">.data</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    foo <span class="op">=</span> <span class="va">bar</span>,</span>
<span>    baz <span class="op">=</span> <span class="va">foo</span><span class="op">^</span><span class="fl">2</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="community-and-support">Community and support<a class="anchor" aria-label="anchor" href="#community-and-support"></a>
</h2>
<p>Please feel free to post questions about <code>biocmask</code>
to:</p>
<ul>
<li>the <a href="https://support.bioconductor.org/" class="external-link">Bioconductor support
site</a>
</li>
<li>as an <a href="https://github.com/jtlandis/biocmask/issues" class="external-link">Issue</a> on
GitHub</li>
<li>in the <code>#tidiness_in_bioc</code> channel on the
<em>community-bioc</em> Slack</li>
</ul>
<p>For code contributions:</p>
<ul>
<li>For small fixes, feel free to post a PR on GitHub</li>
<li>For larger structural changes to the package code, please reach out
to the development team first through one of the above channels.</li>
</ul>
<p>Thanks for your interest in <code>biocmask</code>!</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html" class="external-link">session_info</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">##  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">##  version  R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">##  os       Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">##  system   x86_64, linux-gnu</span></span>
<span><span class="co">##  ui       X11</span></span>
<span><span class="co">##  language en</span></span>
<span><span class="co">##  collate  en_US.UTF-8</span></span>
<span><span class="co">##  ctype    en_US.UTF-8</span></span>
<span><span class="co">##  tz       UTC</span></span>
<span><span class="co">##  date     2024-09-24</span></span>
<span><span class="co">##  pandoc   3.2 @ /usr/bin/ (via rmarkdown)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">##  <span style="color: #555555; font-style: italic;">package             </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version   </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">##  abind                  1.4-8      <span style="color: #555555;">2024-09-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  airway               * 1.24.0     <span style="color: #555555;">2024-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  Biobase              * 2.64.0     <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  BiocGenerics         * 0.50.0     <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  biocmask             * 0.99.12    <span style="color: #555555;">2024-09-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  bslib                  0.8.0      <span style="color: #555555;">2024-07-29</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  cachem                 1.1.0      <span style="color: #555555;">2024-05-16</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  cli                    3.6.3      <span style="color: #555555;">2024-06-21</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  crayon                 1.5.3      <span style="color: #555555;">2024-06-20</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  DelayedArray           0.30.1     <span style="color: #555555;">2024-05-07</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  desc                   1.4.3      <span style="color: #555555;">2023-12-10</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  devtools               2.4.5      <span style="color: #555555;">2022-10-11</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  digest                 0.6.37     <span style="color: #555555;">2024-08-19</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  dplyr                * 1.1.4      <span style="color: #555555;">2023-11-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  ellipsis               0.3.2      <span style="color: #555555;">2021-04-29</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  evaluate               1.0.0      <span style="color: #555555;">2024-09-17</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  fansi                  1.0.6      <span style="color: #555555;">2023-12-08</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  fastmap                1.2.0      <span style="color: #555555;">2024-05-15</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  fs                     1.6.4      <span style="color: #555555;">2024-04-25</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  generics               0.1.3      <span style="color: #555555;">2022-07-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  GenomeInfoDb         * 1.40.1     <span style="color: #555555;">2024-05-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  GenomeInfoDbData       1.2.12     <span style="color: #555555;">2024-06-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  GenomicRanges        * 1.56.1     <span style="color: #555555;">2024-06-12</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  glue                   1.7.0      <span style="color: #555555;">2024-01-09</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  highr                  0.11       <span style="color: #555555;">2024-05-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  htmltools              0.5.8.1    <span style="color: #555555;">2024-04-04</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  htmlwidgets            1.6.4      <span style="color: #555555;">2023-12-06</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  httpuv                 1.6.15     <span style="color: #555555;">2024-03-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  httr                   1.4.7      <span style="color: #555555;">2023-08-15</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  IRanges              * 2.38.1     <span style="color: #555555;">2024-07-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  jquerylib              0.1.4      <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  jsonlite               1.8.9      <span style="color: #555555;">2024-09-20</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  knitr                  1.48       <span style="color: #555555;">2024-07-07</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  later                  1.3.2      <span style="color: #555555;">2023-12-06</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  lattice                0.22-6     <span style="color: #555555;">2024-03-20</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.4.1)</span></span></span>
<span><span class="co">##  lifecycle              1.0.4      <span style="color: #555555;">2023-11-07</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  magrittr               2.0.3      <span style="color: #555555;">2022-03-30</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  Matrix                 1.7-0      <span style="color: #555555;">2024-04-26</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.4.1)</span></span></span>
<span><span class="co">##  MatrixGenerics       * 1.16.0     <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  matrixStats          * 1.4.1      <span style="color: #555555;">2024-09-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  memoise                2.0.1      <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  mime                   0.12       <span style="color: #555555;">2021-09-28</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  miniUI                 0.1.1.1    <span style="color: #555555;">2018-05-18</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  pillar                 1.9.0      <span style="color: #555555;">2023-03-22</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  pkgbuild               1.4.4      <span style="color: #555555;">2024-03-17</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  pkgconfig              2.0.3      <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  pkgdown                2.1.1      <span style="color: #555555;">2024-09-17</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  pkgload                1.4.0      <span style="color: #555555;">2024-06-28</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  profvis                0.4.0      <span style="color: #555555;">2024-09-20</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  promises               1.3.0      <span style="color: #555555;">2024-04-05</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  purrr                  1.0.2      <span style="color: #555555;">2023-08-10</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  R6                     2.5.1      <span style="color: #555555;">2021-08-19</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  ragg                   1.3.3      <span style="color: #555555;">2024-09-11</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  Rcpp                   1.0.13     <span style="color: #555555;">2024-07-17</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  remotes                <span style="color: #BB00BB; font-weight: bold;">2.5.0.9000</span> <span style="color: #555555;">2024-09-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Github (r-lib/remotes@5b7eb08)</span></span></span>
<span><span class="co">##  rlang                  1.1.4      <span style="color: #555555;">2024-06-04</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  rmarkdown              2.28       <span style="color: #555555;">2024-08-17</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  S4Arrays               1.4.1      <span style="color: #555555;">2024-05-20</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  S4Vectors            * 0.42.1     <span style="color: #555555;">2024-07-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  S7                     0.1.1      <span style="color: #555555;">2023-09-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  sass                   0.4.9      <span style="color: #555555;">2024-03-15</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  sessioninfo            1.2.2      <span style="color: #555555;">2021-12-06</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  shiny                  1.9.1      <span style="color: #555555;">2024-08-01</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  SparseArray            1.4.8      <span style="color: #555555;">2024-05-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  SummarizedExperiment * 1.34.0     <span style="color: #555555;">2024-05-01</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  systemfonts            1.1.0      <span style="color: #555555;">2024-05-15</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  textshaping            0.4.0      <span style="color: #555555;">2024-05-24</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  tibble               * 3.2.1      <span style="color: #555555;">2023-03-20</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  tidyr                  1.3.1      <span style="color: #555555;">2024-01-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  tidyselect             1.2.1      <span style="color: #555555;">2024-03-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  UCSC.utils             1.0.0      <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  urlchecker             1.0.1      <span style="color: #555555;">2021-11-30</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  usethis                3.0.0      <span style="color: #555555;">2024-07-29</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  utf8                   1.2.4      <span style="color: #555555;">2023-10-22</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  vctrs                  0.6.5      <span style="color: #555555;">2023-12-01</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  withr                  3.0.1      <span style="color: #555555;">2024-07-31</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  xfun                   0.47       <span style="color: #555555;">2024-08-17</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  xtable                 1.8-4      <span style="color: #555555;">2019-04-21</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  XVector                0.44.0     <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">##  yaml                   2.3.10     <span style="color: #555555;">2024-07-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">##  zlibbioc               1.50.0     <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #555555;"> [1] /__w/_temp/Library</span></span></span>
<span><span class="co">## <span style="color: #555555;"> [2] /usr/local/lib/R/site-library</span></span></span>
<span><span class="co">## <span style="color: #555555;"> [3] /usr/local/lib/R/library</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/jtlandis" class="external-link">Justin Landis</a>, <a href="https://mikelove.github.io" class="external-link">Michael Love</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
